src/interpreter.rs

```diff
@@ -912,139 +989,70 @@ pub fn eval_step<'a>(
                         if stack.size() < 1 {
                             return set_error(ScriptError::InvalidStackOperation);
                         }
-                        let bn = ScriptNum::try_from(stack.top(-1)?.len())
-                            .expect("stack element size fits in ScriptNum");
-                        stack.push_back(bn.getvch())
+                        let bn = i64::try_from(stack.top(-1)?.len())
+                            .expect("stack element size <= MAX_SCRIPT_ELEMENT_SIZE");
+                        stack.push_back(serialize_num(bn))
                     }
 
                     //
                     // Bitwise logic
                     //
-                    OP_EQUAL | OP_EQUALVERIFY => {
-                        // (x1 x2 - bool)
-                        if stack.size() < 2 {
-                            return set_error(ScriptError::InvalidStackOperation);
-                        }
-                        let vch1 = stack.top(-2)?.clone();
-                        let vch2 = stack.top(-1)?.clone();
-                        let equal = vch1 == vch2;
-                        stack.pop()?;
-                        stack.pop()?;
-                        stack.push_back(if equal { VCH_TRUE.to_vec() } else { VCH_FALSE });
-                        if op == OP_EQUALVERIFY {
-                            if equal {
-                                stack.pop()?;
-                            } else {
-                                return set_error(ScriptError::EqualVerify);
-                            }
+                    // (x1 x2 - bool)
+                    OP_EQUAL => binop(stack, |x1, x2| Ok(cast_from_bool(x1 == x2)))?,
```

Comment:
This drops the explicit check for `InvalidStackOperation`. If I understand correctly, it's correct to do that here because `stack.top` will do it, and there are no other side effects / early returns before that could occur.

But there are many many other places where that is the case (and the explicit checks each take 3 lines so they are adding to clutter). Is the plan to remove those soon? If so then no problem, it doesn't need to be in this PR.